Add support for samsung-milletwifi

Susurrus opened the pull request:
Only the following hardware is confirmed working at this point:
 * Hardware buttons
 * display
 * accelerometer
 * USB
 * successful booting

Susurrus commented:
I think [2cb7428](https://github.com/msm8226-mainline/linux/pull/3/commits/2cb74286672c3894cf7e808771d740098420e393) isn't right and breaks some other things. That needs to be cleaned up before this can be merged I think.

wonderfulShrineMaidenOfParadise commented on arch/arm/boot/dts/qcom/qcom-apq8026-samsung-milletwifi.dts:
```diff
@@ -0,0 +1,571 @@
+};
+
+&pronto {
+	status = "okay";
```
Always put `status = "okay";` at the bottom, as the last property.

wonderfulShrineMaidenOfParadise commented on arch/arm/boot/dts/qcom/qcom-apq8026-samsung-milletwifi.dts:
```diff
@@ -0,0 +1,571 @@
+	mx-supply = <&pm8226_l3>;
+	pll-supply = <&pm8226_l8>;
+
+	status = "okay";
```
I don't have a wifi-only or msm8226 device myself so not sure, but are you enabling modem for some reasons?

Susurrus replied:
I was told it might be needed for wifi. There's still a lot unknown with this DTB as whether everything in it is 100% correct. I was told to submit a PR with what I have now tho.

Susurrus replied:
I can leave this out of this PR if that's preferred. Should I only have 100% confirmed working stuff in PRs here? If that's true, then yes, I should remove this.

Susurrus replied:
Touchscreen isn't currently working either, should I remove that as well?

wonderfulShrineMaidenOfParadise commented on arch/arm/boot/dts/qcom/qcom-apq8026-samsung-milletwifi.dts:
```diff
@@ -0,0 +1,571 @@
+		gpio = <&tlmm 31 GPIO_ACTIVE_HIGH>;
+		enable-active-high;
+
+		pinctrl-names = "default";
+		pinctrl-0 = <&tsp_en_default_state>;
```
```suggestion
		pinctrl-0 = <&tsp_en_default_state>;
		pinctrl-names = "default";
```

Suggested by upstream and not really my idea. Could be helpful.

Susurrus replied:
I can reorder this if that's preferred.

Susurrus commented:
Removed the panel code (incl [2cb7428](https://github.com/msm8226-mainline/linux/pull/3/commits/2cb74286672c3894cf7e808771d740098420e393)) as it's not working yet and the commits need some cleanup. Instead going to limit this to other components and simple framebuffer support. I think this can be merged as an initial starting point.

wonderfulShrineMaidenOfParadise commented on arch/arm/boot/dts/qcom/qcom-apq8026-samsung-milletwifi.dts:
```diff
@@ -0,0 +1,571 @@
+		compatible = "st,lis2hh12";
+		reg = <0x1d>;
+
+		interrupt-parent = <&tlmm>;
+		interrupts = <54 IRQ_TYPE_LEVEL_HIGH>;
```
```suggestion
		interrupts-extended = <&tlmm 54 IRQ_TYPE_LEVEL_HIGH>;
```

Another idea from upstream.

Susurrus replied:
Sure, I'll change this.

wonderfulShrineMaidenOfParadise commented on arch/arm/boot/dts/qcom/qcom-apq8026-samsung-milletwifi.dts:
```diff
@@ -0,0 +1,571 @@
+	vqmmc-supply = <&pm8226_l21>;
+
+	bus-width = <4>;
+	cd-gpios = <&tlmm 38 GPIO_ACTIVE_LOW>;
```
It could be better with pinctrl defined?

Susurrus replied:
I'm not certain how to figure out what these should be.

Susurrus replied:
Something like:
```

&sdhc_2 {
	vmmc-supply = <&pm8226_l18>;
	vqmmc-supply = <&pm8226_l21>;

	pinctrl-0 = <&sdc2_default &sdc2_cd_default_state>;
	pinctrl-1 = <&sdc2_sleep &sdc2_cd_default_state>;
	pinctrl-names = "default", "sleep";

	bus-width = <4>;
	cd-gpios = <&tlmm 38 GPIO_ACTIVE_LOW>;

	status = "okay";
};

&tlmm {
	sdc2_cd_default_state: sdc2-cd-default-state {
		pins = "gpio38";
		function = "gpio";
		drive-strength = <2>;
		bias-disable;
	};
```

Is that what you mean? I don't understand what this is doing, but it was referenced in the MSM8916 mainlining doc. From what I see in the downstream DTB I have, it should be correct, but I have no idea how to validate this, either from looking at the hardware or testing the software. Any suggestions?

Susurrus replied:
Compiling this doesn't work as the underlying `sdc2_default` definitions aren't in the msm8226 configs. I can add them, but I don't really understand this subsystem enough to know I'm doing the right thing.

wonderfulShrineMaidenOfParadise replied:
For sure you don't have `sdc2_default`, but `sdhc2_default_state` exists in `qcom-msm8226.dtsi`.

Susurrus replied:
Okay, cool. I think I should at least set the following:
```
	pinctrl-0 = <&sdhc2_default_state>;
	pinctrl-names = "default";
```

There's also `sdhc1_default_state`, should I do that for the `sdhc1` (eMMC) as well?

wonderfulShrineMaidenOfParadise replied:
https://github.com/Susurrus/linux/blob/milletwifi/arch/arm/boot/dts/qcom/qcom-msm8226.dtsi#L349

It's already set on `sdhc1` so there is no need. Here we are trying to add new pinctrl for new introduced cd-gpio, which uses gpio pin 38 on `sdhc2`.

Susurrus replied:
Okay, I'm not following what you're asking me to do here. I don't really understand pinctrl, I'm just copy/pasting from other DTBs. I don't think this is something downstream does, and it seems to be board or platform specific. None of the other apq8026-based devices do this.

From what I saw with the msm8916 devices, they defined some at the platform level, should we be doing the same here with msm8226?

Susurrus replied:
Okay, I think I understood what you were saying and I might have it correct now. PTAL.

Susurrus commented:
Okay, I restricted this to everything that's at least partially working + the hall effect sensor. Not certain why that doesn't work, but it's the same config as matisse-wifi, so I'd like to keep it in here so I can keep playing with it.

Not certain what to do about the pinctrl feedback as well.

Susurrus commented:
I'm actively working on this now, think I may have figured out the lid sensor and doing some other minor cleanup work suggested by the MSM8916 mainlining article.

Susurrus commented:
Okay, the cover sensor now works. Only thing I know that doesn't work correctly is the backlight:
 * It doesn't turn off when the system goes into suspend (like when you close the lid).
 * Manually setting it doesn't change the brightness at all

Looking at the downstream DTB I haven't been able to decipher if there's a pin I'm missing. I'm happy to drop this commit if it moves this along, otherwise some help figuring this out would be appreciated!

wonderfulShrineMaidenOfParadise commented on arch/arm/boot/dts/qcom/qcom-apq8026-samsung-milletwifi.dts:
```diff
@@ -0,0 +1,464 @@
+	bus-width = <4>;
+	cd-gpios = <&tlmm 38 GPIO_ACTIVE_LOW>;
+
+	pinctrl-0 = <&sdc2_cd_default_state>;
```
```suggestion
	pinctrl-0 = <&sdhc2_default_state &sdc2_cd_default_state>;
```

wonderfulShrineMaidenOfParadise replied:
Whoops. Suggested wrong stuffs I guess. Fixed it just now.

z3ntu commented on arch/arm/boot/dts/qcom/qcom-apq8026-samsung-milletwifi.dts:
```diff
@@ -0,0 +1,464 @@
+		};
+	};
+
+	backlight_pwm: pwm {
```
p from pwm comes after i from i2c-backlight, so move the node down

z3ntu commented on arch/arm/boot/dts/qcom/qcom-apq8026-samsung-milletwifi.dts:
```diff
@@ -0,0 +1,464 @@
+	};
+};
+
+&tlmm {
```
move this and rpm_requests up so the nodes are sorted alphabetically

Susurrus replied:
Got it. I followed the approach suggested in the MSM8916 Mainlining article (https://wiki.postmarketos.org/wiki/MSM8916_Mainlining), so I guess that article could use some updates (see the "Device Tree Structure" section). I'll alphabetize things here.

z3ntu replied:
I guess you could argue either way. But I think for msm8226 devices they're all strictly alphabetical

Susurrus commented:
I think the first 5 commits here are pretty uncontroversial and could be merged. I'm still working on the backlight and touchscreen right now. Would it make sense to merge the first 5 now and I'll submit a separate PR for these when they're more solid?

Susurrus commented:
The last commit is pending me getting the firmware files sorted, but the touchscreen is confirmed working now. Known issues:
 * Brightness can't be adjusted. This is driven by PWM from the panel, so needs to wait until then
 * Touchscreen buttons don't work. Will be fixed in 6.6 because of work done to support `samsung-lt01wifi`.

Other systems need to be enables (USB-OTG, battery, charging, wifi/bt, audio, GPS, and irda), but what's already done so far seems solid.

Susurrus commented:
I removed the change with GPS, audio, and wifi/bt as that'll require firmware and additional changes I haven't tested yet.

z3ntu commented on arch/arm/boot/dts/qcom/qcom-apq8026-samsung-milletwifi.dts:
```diff
@@ -0,0 +1,539 @@
+		touchscreen-size-y = <1280>;
+		avdd-supply = <&reg_tsp_3p3v>;
+		vdd-supply = <&reg_tsp_1p8v>;
+		linux,keycodes = <KEY_MENU KEY_BACK>;
```
remove, this is not a valid property for this binding

z3ntu commented on arch/arm/boot/dts/qcom/qcom-apq8026-samsung-milletwifi.dts:
```diff
@@ -0,0 +1,539 @@
+			dev-ctrl = /bits/ 8 <0x80>;
+			init-brt = /bits/ 8 <0x3f>;
+
+			/* Change transition duration: 200ms, Change transition strength: heavy, PWM hysteresis: 1-bit w/ 8-bit resolution */
```
Wrap these comments at 100 characters. For multiline comments use this style
```
/*
 * text here
 */
```

z3ntu commented on arch/arm/boot/dts/qcom/qcom-apq8026-samsung-milletwifi.dts:
```diff
@@ -0,0 +1,539 @@
+	status = "okay";
+
+	touchscreen@48 {
+		/* MELFAS MMS252, using MMS114 compatible for now */
```
Not sure if this will slide upstream, but you can try
Otherwise what you could do is using `compatible = "melfas,mms252", "melfas,mms114";` but then you'd also need to update the bindings to document this

Susurrus replied:
What would be the right way to do this? It seems like the MMS114 driver fully supports the 252 I'm using. Would the right way to do this to modify the mms114 driver to support the mms252 "officially"?

z3ntu replied:
If it doesn't need any driver changes then `compatible = "melfas,mms252", "melfas,mms114";` + dt bindings update should be the correct way (also fine if for now you can tell "it generally works")

Susurrus replied:
Didn't know what to do with this one, so left it as-is. Looked at modifying the bindings to allow for `compatible = "melfas,mms252"` and it seems more than trivial, so hoping we can slide this by like this.

z3ntu replied:
This should work for this use case. Feel free to put your author into the commit
```diff
diff --git a/Documentation/devicetree/bindings/input/touchscreen/melfas,mms114.yaml b/Documentation/devicetree/bindings/input/touchscreen/melfas,mms114.yaml
index fdd02898e249..b39724e92c29 100644
--- a/Documentation/devicetree/bindings/input/touchscreen/melfas,mms114.yaml
+++ b/Documentation/devicetree/bindings/input/touchscreen/melfas,mms114.yaml
@@ -17,13 +17,17 @@ properties:
     pattern: "^touchscreen(@.*)?$"
 
   compatible:
-    items:
+    oneOf:
       - enum:
           - melfas,mms114
           - melfas,mms134s
           - melfas,mms136
           - melfas,mms152
           - melfas,mms345l
+      - items:
+          - enum:
+              - melfas,mms252
+          - const: melfas,mms114
 
   reg:
     description: I2C address

```

z3ntu replied:
plus this
```diff
diff --git a/arch/arm/boot/dts/qcom/qcom-apq8026-samsung-milletwifi.dts b/arch/arm/boot/dts/qcom/qcom-apq8026-samsung-milletwifi.dts
index 06855c70a12d..6bee5cc12503 100644
--- a/arch/arm/boot/dts/qcom/qcom-apq8026-samsung-milletwifi.dts
+++ b/arch/arm/boot/dts/qcom/qcom-apq8026-samsung-milletwifi.dts
@@ -276,8 +276,7 @@ &blsp1_i2c5 {
 	status = "okay";
 
 	touchscreen@48 {
-		/* MELFAS MMS252, using MMS114 compatible for now */
-		compatible = "melfas,mms114";
+		compatible = "melfas,mms252", "melfas,mms114";
 		reg = <0x48>;
 		interrupts-extended = <&tlmm 17 IRQ_TYPE_EDGE_FALLING>;
 		touchscreen-size-x = <800>;

```

z3ntu commented on arch/arm/boot/dts/qcom/qcom-apq8026-samsung-milletwifi.dts:
```diff
@@ -0,0 +1,539 @@
+	bus-width = <4>;
+	cd-gpios = <&tlmm 38 GPIO_ACTIVE_LOW>;
+
+	pinctrl-0 = <&sdhc2_default_state &sdc2_cd_default_state>;
```
`pinctrl-0 = <&sdhc2_default_state>, <&sdc2_cd_default_state>;`

z3ntu commented:
Looks pretty good!

Now last thing for now (I hope) is getting the commit messages right.

The commit adding dts file (and the Makefile entry) should be the last commit in the series and titled something like `ARM: dts: qcom: Add support for Samsung Galaxy Tab 4 8.0 Wi-Fi` with a decent commit message what you're doing.

For the format of commit messages you can always check `git log <path-to-file>` (or if you're adding a new file, check that directory for examples).

Every commit needs to have `Signed-off-by: Your Name <your@email.com>`: https://www.kernel.org/doc/html/latest/process/submitting-patches.html#sign-your-work-the-developer-s-certificate-of-origin

And your patchset should be two patches in total: one adding the string to `Documentation/devicetree/bindings/arm/qcom.yaml` (also check the commit messages example there), the other one adding the dts (+makefile change). No need to have them separate for initial bringup.

All the patches that come later should be one per feature for example though.

Susurrus commented:
> And your patchset should be two patches in total: one adding the string to Documentation/devicetree/bindings/arm/qcom.yaml (also check the commit messages example there), the other one adding the dts (+makefile change). No need to have them separate for initial bringup.
>
> All the patches that come later should be one per feature for example though.

Does "initial bringup" mean everything I've included in this PR? My interpretation here is that there should be 2 commits as you mentioned. Then, in subsequent PRs I'll file, I should do commits per-feature (much like I did already in this initial PR).

Is there a reason you want me to squash all the DTS changes into 1 commit rather than leave them as single commits per-feature as I've already done? I'd think that'd be easiest to review and clearest in the git history. And since you mention it's what I should do for subsequent feature work, seems most consistent.

z3ntu commented:
> Is there a reason you want me to squash all the DTS changes into 1 commit rather than leave them as single commits per-feature as I've already done? I'd think that'd be easiest to review and clearest in the git history. And since you mention it's what I should do for subsequent feature work, seems most consistent.

Upstream doesn't want a thousand commits for initial dts inclusion :)

But yes I mean all the dts commit here you squash into one which adds the features here at once.

Susurrus committed:
dt-bindings: arm: qcom: Document samsung,milletwifi device

Add binding documentation for Samsung Galaxy Tab 4 8.0 Wi-Fi
tablet which is based on Snapdragon 400 (apq8026) SoC.

Signed-off-by: Bryant Mairs <bryant@mai.rs>

Susurrus committed:
ARM: dts: qcom: Add support for Samsung Galaxy Tab 4 8.0 Wi-Fi

Add support for this tablet based on the MSM8226 SoC, codenamed
"millet".

Signed-off-by: Bryant Mairs <bryant@mai.rs>

Susurrus commented:
Okay, should be ready! I'm compiling it now to push it to the device to make sure I didn't mess anything up.

Susurrus commented:
Alright, tested as working on my device. I think we're gtg here! :-)

z3ntu commented:
@Susurrus you wanna add the `melfas,mms252` change still? https://github.com/msm8226-mainline/linux/pull/3#discussion_r1362594376

Otherwise I'll add it before merging :)

Susurrus commented:
Feel free to add it. Wasn't certain what you meant and won't be able to do it tonight at this point.

On October 17, 2023 9:17:35 PM GMT+02:00, Luca Weiss ***@***.***> wrote:
>@Susurrus you wanna add the `melfas,mms252` change still? https://github.com/msm8226-mainline/linux/pull/3#discussion_r1362594376
>
>Otherwise I'll add it before merging :)
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/msm8226-mainline/linux/pull/3#issuecomment-1767018440
>You are receiving this because you were mentioned.
>
>Message ID: ***@***.***>

z3ntu committed:
dt-bindings: input: melfas,mms114: add MMS252 compatible

Add a compatible for MMS252 touchscreen which appears to work fine with
the MMS114 driver.

Signed-off-by: Luca Weiss <luca@z3ntu.xyz>

z3ntu commented:
okay, modified https://github.com/msm8226-mainline/linux/pull/3#discussion_r1362594376 and https://github.com/msm8226-mainline/linux/pull/3#discussion_r1362753329

LGTM now!!
